### DISCUSSION TITLE
Python Package isn't downloaded

### DISCUSSION THREAD

**Discussion by netsonicyxf** [Tunny-gh/Tunny_in_Open_Sea#183] [General]
Win10 x64, Rhino 7.30, Tunny 0.80
After double click Tunny component, it shows this window and I wait for ever
![image](https://github.com/hrntsm/Tunny/assets/43513391/828c5a70-24ca-4b80-b0b4-3f55398db9af)

There is only one new file called "Settings.json" in the 
C:\Users\user name\AppData\Roaming\Grasshopper\Libraries\Tunny
The python package is never be download.

Could you please provide offline python package?

**Comment by hrntsm** [comment]
Hi @netsonicyxf 

I've created a modified version, could you try this one? This one contains the Python Package inside.
https://github.com/hrntsm/Tunny/releases/tag/v0.8.1

**Comment by netsonicyxf** [comment]
Thanks.
When I run built-in "Karamba3d_column_position_opt.gh", it shows 
![image](https://github.com/hrntsm/Tunny/assets/43513391/c133cc3e-8680-4379-9fba-bd889b5dee0c)

**Reply by hrntsm** [reply]
First of all, I'm glad you didn't have any Python problems.
There is a Storage setting in the File tab of Tunny, so press "set file path" button and specify SQLite storage as the file format, and I think it should work.

<img width="922" alt="Screenshot 2023-07-30 at 20 59 11" src="https://github.com/hrntsm/Tunny/assets/23289252/be46b3df-f746-4752-a01d-7ae8028c948b">

**Reply by netsonicyxf** [reply]
Thanks for the tip.

I saved the SQLite file here
![image](https://github.com/hrntsm/Tunny/assets/43513391/0aa60427-8a86-49c9-a00e-f1e0b685cbd6)

When running Tunny, it shows "调用的目标发生了异常。", which means Exception Has Been Thrown by the Target of an Invocation
![image](https://github.com/hrntsm/Tunny/assets/43513391/9721e2aa-bbec-4c3a-8ebf-fb7b1e4d878a)

**Reply by hrntsm** [reply]
Is this what you see when you run "Karamba3d_column_position_opt.gh"? And is it the same when "sample.gh" is executed?

**Reply by netsonicyxf** [reply]
1. I saw the above error everytime when running "Karamba3d_column_position_opt.gh" or "Vase_Shape_opt.gh", which asked me to specify SQLite storage as the file format. Do I have to use .sqlite for examples "Karamba3d_column_position_opt.gh" or "Vase_Shape_opt.gh"
I tried to create fish.sqlite, but Tunny didn't create anything. But If create fish.db, it will show the above error
2. I didn't see the error when running "sample.gh" yesterday. But today I saw it once. I exited Rhino and then reopened "sample.gh" and ran it, it worked, no error occured.

**Reply by hrntsm** [reply]
Human-in-the-loop optimization only supports files in sqlite format. Other optimizations support both sqlite and journal format, so either is fine.

**Reply by netsonicyxf** [reply]
I did specify SQLite storage as the file format, but Tunny never created any .sqlite file for me.
I think that's why Human-in-the-loop optimization doesn't work, 
Any solution?

**Reply by hrntsm** [reply]
It's working in my environment, so I can't identify a solution with the current information.
If the db file was created from Tunny's UI and it doesn't work well with optimization afterwards, please provide the db file.

**Reply by netsonicyxf** [reply]
Maybe I wasn't being clear.
I use Vase_Shape_opt.gh as example for Human-in-the-loop,
1. If I save output file as Vase_Shape_opt.db (in attachment) then run optimization, it shows error
![image](https://github.com/hrntsm/Tunny/assets/43513391/ad806ec2-f1e5-47f9-beb5-353663ef057e)
[Vase_Shape_opt.zip](https://github.com/hrntsm/Tunny/files/12265893/Vase_Shape_opt.zip)

2. If I save output file as Vase_Shape_opt.sqlite, Tunny never create it.

**Reply by hrntsm** [reply]
I checked the file you gave me and it was created properly as a sqlite file.
Possible problems are as follows.
When Human-in-the-loop is run, it automatically creates a folder to store the results, but if the directory does not have permissions to create new files, it may not work.
The Human-in-the-loop files are created in the same folder as the db file.

**Reply by netsonicyxf** [reply]
The sample.gh and Vase_Shape_opt.gh are in the same folder.
The sample.gh can be run without any problem, which means the directory do have permissions to create new files.

**Reply by hrntsm** [reply]
It doesn't matter where the gh file is, it's where the db file is. db files are created in a place where it doesn't matter to create new files as well?

**Reply by netsonicyxf** [reply]
I mean the db file can be created for sample.gh, which means the directory do have permissions to create new files. 
Since Vase_Shape_opt.gh and sample.gh are in the same folder, Human-in-the-loop can automatically creates a folder to store the results in it.

**Reply by hrntsm** [reply]
Thanks for the response.
I do not know how to fix the current situation. I am currently working on a support for the latest version of optuna. I will let you know when it is done, as I plan to improve the ease of use of human-in-the-loop with this support.

**Comment by netsonicyxf** [comment]
I have some question about sample.gh. Everytime I ran it, I got different answer
1st time
![image](https://github.com/hrntsm/Tunny/assets/43513391/5f0f679f-12c8-42e9-9230-5baf13a9bdc2)

2nd time
![image](https://github.com/hrntsm/Tunny/assets/43513391/027e5aac-d5f0-41cf-b0ad-6c8d81632587)

Are they suppoed to be the same?

**Reply by hrntsm** [reply]
All optimizations handled by Tunny involve randomness in intermediate steps of the algorithm. Therefore, unless you directly specify the seed value of the random, the result of the optimization will change each time.

Also, sample.gh performs multi-objective optimization. In the case of multi-objective optimization, multiple solutions (Pareto solution) are obtained instead of a single optimal solution, so the optimization results are not reflected in Grasshopper's slider. The results can be reflected on the Grasshopper side by specifying the necessary values from the Output tab.

Check the following for seed value settings.
- https://github.com/hrntsm/Tunny/discussions/169

For information on how to check and output results, see below
- https://tunny-docs.deno.dev/docs/optimize-window/visualize-tab
- https://tunny-docs.deno.dev/docs/optimize-window/output-tab
